src/protocols/bdx/BdxMessages.cpp

```diff
@@ -0,0 +1,456 @@
+    uint8_t * bufStart      = aBuffer.Start();
+    Reader bufReader(bufStart, aBuffer.DataLength());
+    BitFlags<uint8_t, RangeControlFlags> rangeCtlFlags;
+
+    SuccessOrExit(bufReader.Read8(&proposedTransferCtl).Read8(&rangeCtl).Read16(&MaxBlockSize).StatusCode());
+
+    Version = proposedTransferCtl & kVersionMask;
+    TransferCtlOptions.SetRaw(static_cast<uint8_t>(proposedTransferCtl & ~kVersionMask));
+    rangeCtlFlags.SetRaw(rangeCtl);
+
+    if (rangeCtlFlags.Has(kStartOffset))
+    {
+        if (rangeCtlFlags.Has(kWiderange))
+        {
+            SuccessOrExit(bufReader.Read64(&StartOffset).StatusCode());
+        }
+        else
+        {
+            SuccessOrExit(bufReader.Read32(&tmpUint32Value).StatusCode());
+            StartOffset = tmpUint32Value;
+        }
+    }
+
+    if (rangeCtlFlags.Has(kDefLen))
+    {
+        if (rangeCtlFlags.Has(kWiderange))
+        {
+            SuccessOrExit(bufReader.Read64(&MaxLength).StatusCode());
+        }
+        else
+        {
+            SuccessOrExit(bufReader.Read32(&tmpUint32Value).StatusCode());
+            MaxLength = tmpUint32Value;
+        }
+    }
+
+    SuccessOrExit(bufReader.Read16(&FileDesLength).StatusCode());
+
+    // WARNING: this struct will store a pointer to the start of the file designator in the PacketBuffer,
+    // but will not make a copy. It is essential that this struct not outlive the PacketBuffer,
+    // or there is risk of unsafe memory access.
+    VerifyOrExit(bufReader.HasAtLeast(FileDesLength), err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+    FileDesignator = &bufStart[bufReader.OctetsRead()];
+
+    // Rest of message is metadata (could be empty)
+    if (bufReader.Remaining() > FileDesLength)
+    {
+        // WARNING: this struct will store a pointer to the start of metadata in the PacketBuffer,
+        // but will not make a copy. It is essential that this struct not outlive the PacketBuffer,
+        // or there is risk of unsafe memory access.
+        uint16_t metadataStartIndex = static_cast<uint16_t>(bufReader.OctetsRead() + FileDesLength);
+        Metadata                    = &bufStart[metadataStartIndex];
+        MetadataLength              = static_cast<uint16_t>(aBuffer.DataLength() - metadataStartIndex);
+    }
+
+exit:
+    if (bufReader.StatusCode() != CHIP_NO_ERROR)
+    {
+        err = bufReader.StatusCode();
+    }
+    return err;
+}
+
+size_t TransferInit::PackedSize() const
+{
+    System::PacketBuffer * pbuf = System::PacketBuffer::NewWithAvailableSize(0);
+    BufBound emptyBuf(pbuf->Start(), 0);
+    Pack(emptyBuf);
+    return emptyBuf.Written();
+}
+
+bool TransferInit::operator==(const TransferInit & another) const
+{
+    if (MetadataLength != another.MetadataLength || FileDesLength != another.FileDesLength)
+    {
+        return false;
+    }
+    bool fileDesMatches  = memcmp(FileDesignator, another.FileDesignator, FileDesLength) == 0;
```

Comment:
These memcmp should only occur if the respective length > 0

Reply:
ack